crypto/vault/vault.go

```diff
@@ -0,0 +1,158 @@
+// vault.go - cryptographic vault for mixnet client key material
+// Copyright (C) 2017  David Anthony Stainton
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+// crypto vault
+package vault
+
+import (
+	"bytes"
+	"crypto/rand"
+	"encoding/pem"
+	"errors"
+	"io/ioutil"
+	"os"
+
+	"github.com/magical/argon2"
+	"golang.org/x/crypto/nacl/secretbox"
+)
+
+const (
+	// argon2SaltSize is the salt size for use with argon2
+	argon2SaltSize = 8
+
+	// passphraseMinSize is the minimum allowed passphrase size
+	passphraseMinSize = 12
+
+	// secretboxNonceSize is the nonce size for NaCl SecretBox
+	secretboxNonceSize = 24
+)
+
+// Vault is used to Encrypt sensitive data to disk.
+// Uses argon2 for keystretching and NaCl SecretBox
+// for encryption.
+type Vault struct {
+	Type       string
+	Passphrase string
+	Path       string
+	Email      string
+}
+
+// New creates a new Vault
+func New(vaultType, passphrase, path, email string) (*Vault, error) {
+	if len(passphrase) < passphraseMinSize {
+		return nil, errors.New("passphrase too short")
+	}
+	v := Vault{
+		Type:       vaultType,
+		Email:      email,
+		Passphrase: passphrase,
+		Path:       path,
+	}
+	return &v, nil
+}
+
+// stretch performs argon2 key stretching on the given passphrase
+func (v *Vault) stretch(passphrase string) ([]byte, error) {
+	salt := passphrase[0:argon2SaltSize]
+	pass := passphrase[argon2SaltSize:]
+	par := 2
+	mem := int64(1 << 16)
+	keyLen := 32
+	n := 32
+	out, err := argon2.Key([]byte(pass), []byte(salt), n, par, mem, keyLen)
+	if err != nil {
+		return nil, err
+	}
+	return out, nil
+}
+
+// Open returns decrypted data from the vault
+func (v *Vault) Open() ([]byte, error) {
+	pemPayload, err := ioutil.ReadFile(v.Path)
+	if err != nil {
+		return nil, err
+	}
+	block, _ := pem.Decode(pemPayload)
+	if block == nil {
+		return nil, errors.New("failed to decode pem file")
+	}
+	if len(block.Headers) == 0 {
```

Comment:
block.Headers is not used. why is it checked?

Reply:
it *is* used... we use PEM headers in this api when we encode the key and write it to disk; PEM header contains account e-mail.

Reply:
hmm but yeah we could skip this check. not strictly needed